usnistgov/oar-auth-py
# Python Flask-based Authentication Service

This PR completes the implementation of the OAR Authentication Broker service built on the `python3-saml` library intended to replace the Java-based service provided in oar-auth-api.  In addition, this PR adds an implementation of an Identity Provider (IDP) which can be used in testing and development contexts in lieu of the NIST IDP login service.  This repo provides docker containers for both services that allow one to run each standalone. 

This PR calls for two key tests: _(1)_ testing the Authentication service against the NIST SAML-based login service (specifically by @deoyani), and _(2)_ testing the service against the IDP included with this repo.  

### Testing against the NIST login service

This test has been designed specifically for @deoyani.  To test:

1. Check out the `test/saml-sp` branch of this repository
2. Build the docker containers by typing `docker/dockbuild.sh`
3. Change into the `docker/authserver` directory
4. Start the service with a configuration pointing to the NIST login service:
```
    run.sh -c deoyani-authservice-config.yml -p 8000 --bg 
```
This starts the service (in the background) and makes it available at https://localhost:8000/sso/.
5. Ensure that the service's SAML metadata is available at https://localhost:8000/sso/metadata/
6. Exercise the login process by visiting https://localhost:8000/sso/saml/login?redirectTo=https://localhost:8000/sso/_tokeninfo.  
7. At the end of the sequence, ensure that you see in the browser a JSON object contain your user information and a token.  

To stop the server, type `run.sh stop`

### Testing the services standalone

Using the same branch, `test/saml-sp`, test the combination of the Authentication Broker service connected to a locally running IDP/login service with the following:

1. Change into `docker/authserver` directory
2. If you haven't already built the docker containers, do so by typing, `docker/dockbuild.sh`
3. Start both servers (in the background) with the following:
```
    run.sh -I --bg
```
This starts the service (in the background) and makes it available at https://localhost:9095/sso/; the IDP service will be running at https://localhost:8088/.
4. Ensure that the authentication service's SAML metadata is available at https://localhost:9095/sso/metadata/
5. Ensure that the IDP/login service's SAML metadata is available at https://localhost:8088/idp.xml.
6. Test the login sequence by visiting https://localhost:9095/sso/saml/login?redirectTo=https://localhost:9095/sso/_tokeninfo.  
7. When you see the login page, use "upper" as the user name, and "crust" as the password.  (Other configured usernames and passwords can be found in [idp/idp_users.py](https://github.com/usnistgov/oar-auth-py/blob/int/saml-sp/idp/idp_user.py)).
8. At the end of the sequence, ensure that you see in the browser a JSON object contain user information and a token.  

To stop both servers, type `run.sh -I stop`
